
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>funkea &#8212; funkea 0.1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorials" href="tutorials/index.html" />
    <link rel="prev" title="funkea" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="funkea">
<h1>funkea<a class="headerlink" href="#funkea" title="Permalink to this heading">¶</a></h1>
<p>Perform functional enrichment analysis at scale.</p>
<section id="install">
<h2>Install<a class="headerlink" href="#install" title="Permalink to this heading">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">funkea</span></code> is available on PyPI, and can be installed using <code class="docutils literal notranslate"><span class="pre">pip</span></code>:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>funkea
</pre></div>
</div>
<p>If you want to build from source, you can clone the repository and install using:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>git<span class="w"> </span>clone<span class="w"> </span>...
<span class="nb">cd</span><span class="w"> </span>funkea
pip<span class="w"> </span>install<span class="w"> </span>.
</pre></div>
</div>
<p>Note that <code class="docutils literal notranslate"><span class="pre">funkea</span></code> requires Python 3.10 or higher, and that it uses Scala for some of its
user-defined functions. The above command will fail if you do not have a Scala compiler installed,
or if it is not on your path. If you do not have a Scala compiler installed, you can install one
from <a class="reference external" href="https://www.scala-lang.org/download/">here</a> (version 2.12.x). Note that the Scala installer
can sometimes have issues adding the compiler to your path, so you may need to do this manually.</p>
</section>
<section id="quickstart">
<h2>Quickstart<a class="headerlink" href="#quickstart" title="Permalink to this heading">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">funkea</span></code> was built with composability in mind, but also ships with 5 popular enrichment methods out
of the box. The simplest and fastest method is a Fisher’s exact test on the overlapped annotations:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">funkea.implementations</span> <span class="kn">import</span> <span class="n">Fisher</span>
<span class="kn">from</span> <span class="nn">pyspark.sql</span> <span class="kn">import</span> <span class="n">DataFrame</span>

<span class="n">sumstats</span><span class="p">:</span> <span class="n">DataFrame</span> <span class="o">=</span> <span class="o">...</span>  <span class="c1"># your GWAS sumstats</span>

<span class="c1"># perform tissue enrichment using GTEx data</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">Fisher</span><span class="o">.</span><span class="n">default</span><span class="p">()</span>
<span class="n">enrichment</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">sumstats</span><span class="p">)</span>
</pre></div>
</div>
<p>This assumes that the default filepaths are set in the <span class="xref myst">file-registry</span>.
If you have not set the file-registry, you can pass the filepaths directly to the data components:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">funkea.implementations</span> <span class="kn">import</span> <span class="n">Fisher</span>
<span class="kn">from</span> <span class="nn">funkea.core</span> <span class="kn">import</span> <span class="n">data</span>
<span class="kn">from</span> <span class="nn">pyspark.sql</span> <span class="kn">import</span> <span class="n">DataFrame</span>

<span class="n">sumstats</span><span class="p">:</span> <span class="n">DataFrame</span> <span class="o">=</span> <span class="o">...</span>  <span class="c1"># your GWAS sumstats</span>

<span class="c1"># define your annotation component</span>
<span class="n">annotation</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">AnnotationComponent</span><span class="p">(</span>
    <span class="c1"># the column names in annotation dataframe</span>
    <span class="n">columns</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">AnnotationColumns</span><span class="p">(</span>
        <span class="n">annotation_id</span><span class="o">=</span><span class="s2">&quot;gene_id&quot;</span><span class="p">,</span>
        <span class="n">partition_id</span><span class="o">=</span><span class="s2">&quot;tissue_id&quot;</span><span class="p">,</span>
    <span class="p">),</span>
    <span class="c1"># the path to the GTEx dataset (needs to be in parquet format)</span>
    <span class="n">dataset</span><span class="o">=</span><span class="s2">&quot;path/to/gtex_dataset.parquet&quot;</span><span class="p">,</span>
    <span class="c1"># the type of partitioning used in the annotation dataset</span>
    <span class="c1"># (either data.PartitionType.SOFT or data.PartitionType.HARD)</span>
    <span class="c1"># see docs for more information</span>
    <span class="n">partition_type</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">PartitionType</span><span class="o">.</span><span class="n">SOFT</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># perform tissue enrichment using GTEx data</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">Fisher</span><span class="o">.</span><span class="n">default</span><span class="p">(</span><span class="n">annotation</span><span class="o">=</span><span class="n">annotation</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">dataset</span></code> can also be passed as a <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code> directly, if you have already loaded it into your
Spark session.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">funkea.core</span> <span class="kn">import</span> <span class="n">data</span>
<span class="kn">from</span> <span class="nn">pyspark.sql</span> <span class="kn">import</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">SparkSession</span>

<span class="c1"># load the GTEx dataset</span>
<span class="n">spark</span><span class="p">:</span> <span class="n">SparkSession</span> <span class="o">=</span> <span class="o">...</span>  <span class="c1"># your Spark session</span>
<span class="n">gtex</span><span class="p">:</span> <span class="n">DataFrame</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">parquet</span><span class="p">(</span><span class="s2">&quot;path/to/gtex_dataset.parquet&quot;</span><span class="p">)</span>

<span class="c1"># define your annotation component</span>
<span class="n">annotation</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">AnnotationComponent</span><span class="p">(</span>
    <span class="n">columns</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">AnnotationColumns</span><span class="p">(</span>
        <span class="n">annotation_id</span><span class="o">=</span><span class="s2">&quot;gene_id&quot;</span><span class="p">,</span>
        <span class="n">partition_id</span><span class="o">=</span><span class="s2">&quot;tissue_id&quot;</span><span class="p">,</span>
    <span class="p">),</span>
    <span class="c1"># the GTEx dataset</span>
    <span class="n">dataset</span><span class="o">=</span><span class="n">gtex</span><span class="p">,</span>  <span class="c1"># &lt;-- pass the DataFrame directly</span>
    <span class="n">partition_type</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">PartitionType</span><span class="o">.</span><span class="n">SOFT</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<p>It is generally recommended to set the file-paths using the
<span class="xref myst">file-registry</span>, as it is easy to forget to pass the filepaths to all
the components. The above example is simple, where there is only one required data source (barring
the GWAS sumstats), but more complex methods may require multiple data sources. For example, the
<code class="docutils literal notranslate"><span class="pre">GARFIELD</span></code> method requires linkage disequilibrium (LD) estimates, variant-level controlling
covariates, and the annotation component.</p>
</section>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">funkea</span></code> is a Python library for large-scale functional enrichment analysis. It provides 5 popular
enrichment methods, and also allows for experimentation by composing different components. It is
written in Spark, and allows users to run an arbitrary number of GWAS studies concurrently, given
the resources are available. It also provides a CLI for managing data sources.</p>
<p><code class="docutils literal notranslate"><span class="pre">funkea</span></code> has a few concepts used for abstraction, such that all methods could be unified. A view of
the schematic is outlined below</p>
<p><img alt="schematic" src="_images/schematic.png" /></p>
<p>i.e. each workflow consists of (1) a data pipeline; and (2) an enrichment method. The former filters
down the sumstats (<code class="docutils literal notranslate"><span class="pre">variant_selection</span></code>), creates loci from the remaining variants
(<code class="docutils literal notranslate"><span class="pre">locus_definition</span></code>) and then finally associates these loci with annotations (<code class="docutils literal notranslate"><span class="pre">annotation</span></code>). The
latter then takes the loci (including their annotations) and computes the study-wide enrichments
for each annotation partition, and its respective significance.</p>
<p>The variant selection and locus definitions are composed by the user, but each of the enrichment
methods provided by <code class="docutils literal notranslate"><span class="pre">funkea</span></code> provide default configurations. The user can also define their own
annotation component, which is required for all enrichment methods.</p>
</section>
<section id="setting-default-filepaths">
<h2>Setting default filepaths<a class="headerlink" href="#setting-default-filepaths" title="Permalink to this heading">¶</a></h2>
<p>For ease of use, <code class="docutils literal notranslate"><span class="pre">funkea</span></code> uses a file-registry for its source of truth of various data sources. These
need to be set by a user, which can be set easily by using the <code class="docutils literal notranslate"><span class="pre">funkea-cli</span></code>. For example, if the
user has put all the data into a single directory, the registry can be set like so:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>funkea-cli<span class="w"> </span>init<span class="w"> </span>--from-stem<span class="w"> </span>&lt;PATH_TO_DATASET&gt;
</pre></div>
</div>
<p>This will set the default filepaths for all the data sources, where the filenames will be the key
in JSON object. For example, if the user has put all the data into <code class="docutils literal notranslate"><span class="pre">/path/to/data</span></code>, the registry
will look like:</p>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
<span class="w">  </span><span class="nt">&quot;gtex&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/data/gtex&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="nt">&quot;ld_reference_data&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/data/ld_reference_data&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="nt">&quot;depict_null_loci&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/data/depict_null_loci&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="nt">&quot;chromosomes&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/data/chromosomes&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="nt">&quot;snpsea_background_variants&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/data/snpsea_background_variants&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="nt">&quot;garfield_control_covariates&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/data/garfield_control_covariates&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="nt">&quot;ldsc_controlling_ld_scores&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/data/ldsc_controlling_ld_scores&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="nt">&quot;ldsc_weighting_ld_scores&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/data/ldsc_weighting_ld_scores&quot;</span>
<span class="p">}</span>
</pre></div>
</div>
<p>The filepaths can also be set individually using the <code class="docutils literal notranslate"><span class="pre">funkea-cli</span></code>:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>funkea-cli<span class="w"> </span>init
</pre></div>
</div>
<p>This will start an interactive prompt, where the user can set the filepaths individually. The
registry can also be edited manually at
<code class="docutils literal notranslate"><span class="pre">funkea/core/resources/file_registry.json</span></code>, but it is recommended to use the CLI, as it can be
difficult to find the registry file.</p>
<p>Alternatively, the file paths can be specified directly in the components
(see <span class="xref myst">quickstart</span>). It is generally recommended to use the CLI, as it is easy to forget
to pass the filepaths to all the components, especially for enrichment methods which require
multiple data sources (LD data, controlling covariates etc.).</p>
</section>
<section id="testing">
<h2>Testing<a class="headerlink" href="#testing" title="Permalink to this heading">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">funkea</span></code> uses <code class="docutils literal notranslate"><span class="pre">pytest</span></code> for testing. To run the tests, run the following command from the root
directory:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>make<span class="w"> </span><span class="nb">test</span>
</pre></div>
</div>
<p>This will install most dependencies if they are not present, except the Scala compiler, as this
depends heavily on the user’s system. Make sure to install this yourself first before running the
tests. You can install them from <a class="reference external" href="https://www.scala-lang.org/download/">here</a>.</p>
</section>
<section id="documentation">
<h2>Documentation<a class="headerlink" href="#documentation" title="Permalink to this heading">¶</a></h2>
<p>The documentation is hosted on <a class="reference external" href="https://funkea.readthedocs.io/en/latest/">readthedocs</a>. It is
automatically built from the <code class="docutils literal notranslate"><span class="pre">docs</span></code> directory in the <code class="docutils literal notranslate"><span class="pre">main</span></code> branch. To build the documentation
locally, run the following command from the root directory:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>make<span class="w"> </span>docs
</pre></div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">funkea</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">User Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#install">Install</a></li>
<li class="toctree-l2"><a class="reference internal" href="#quickstart">Quickstart</a></li>
<li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#setting-default-filepaths">Setting default filepaths</a></li>
<li class="toctree-l2"><a class="reference internal" href="#testing">Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#documentation">Documentation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference/index.html">API Reference</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">funkea</a></li>
      <li>Next: <a href="tutorials/index.html" title="next chapter">Tutorials</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2023-2023, Benjamin Tenmann.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="_sources/user_guide.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>