
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>GARFIELD &#8212; funkea 0.1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="LDSC" href="implementations.ldsc.html" />
    <link rel="prev" title="Fisher" href="implementations.fisher.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="module-funkea.implementations.garfield">
<span id="garfield"></span><h1>GARFIELD<a class="headerlink" href="#module-funkea.implementations.garfield" title="Permalink to this heading">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.garfield.ControllingCovariates">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.garfield.</span></span><span class="sig-name descname"><span class="pre">ControllingCovariates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">continuous_covariates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">categorical_covariates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">('binned_tss_distance',</span> <span class="pre">'binned_n_proxies')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.garfield.ControllingCovariates" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads variant-level controlling covariates to include in the regression.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>continuous_covariates</strong> – A sequence of column names, which define controlling covariates with
numerical values. (<code class="docutils literal notranslate"><span class="pre">default=()</span></code>)</p></li>
<li><p><strong>categorical_covariates</strong> – A sequence of column names, which define the controlling covariates
with categorical values. (<code class="docutils literal notranslate"><span class="pre">default=(&quot;binned_tss_distance&quot;,</span> <span class="pre">&quot;binned_n_proxies&quot;)</span></code>)</p></li>
<li><p><strong>dataset</strong> – Path to the controlling covariates, or the dataframe itself.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.garfield.GARFIELD">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.garfield.</span></span><span class="sig-name descname"><span class="pre">GARFIELD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pipeline</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Pipeline</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Method</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.garfield.GARFIELD" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes functional enrichment using <code class="docutils literal notranslate"><span class="pre">GARFIELD</span></code>.</p>
<p><code class="docutils literal notranslate"><span class="pre">GARFIELD</span></code> is a general method for functional enrichment analysis <a class="footnote-reference brackets" href="#id2" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>. It regresses locus
association to a given trait on whether the lead variant — or one of its LD proxies —
overlaps a given partition <span class="math notranslate nohighlight">\(i\)</span>, as well as a set of controlling covariates. Formally:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\mathbb{E} \left[ \mathbb{1} \left[ \boldsymbol{\chi}^2 &gt; \theta \right] | \mathbf{C}, \mathbf{K} \right] &amp;= \left(1 + \exp \left(- \mathbf{z}  - b \right) \right)^{\odot -1} \\
\mathbf{z} &amp;= \mathbf{C}^T \boldsymbol{\alpha} + \mathbf{K}^T \boldsymbol{\beta}\end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(\boldsymbol{\chi}^2 \in \mathbb{R}^M\)</span> are the association statistics for the LD
pruned lead variants, <span class="math notranslate nohighlight">\(\theta\)</span> is some threshold,
<span class="math notranslate nohighlight">\(\mathbf{C} \in \{0, 1\}^{D \times M}\)</span> are the controlling covariates and
<span class="math notranslate nohighlight">\(\mathbf{K}\)</span> is:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\mathbf{K} &amp;= t \left( \mathbf{Q} \right) \\
\mathbf{Q} &amp;= \mathbf{X} \mathbf{L} \\
t(Q_{il}) &amp;= \mathbb{1} \left[ Q_{il} &gt; 0 \right]\end{split}\]</div>
<p>an indicator matrix showing whether a given locus overlaps an annotation partition. The
definitions of <span class="math notranslate nohighlight">\(\mathbf{X}\)</span> and <span class="math notranslate nohighlight">\(\mathbf{L}\)</span> are the same here as for
<a class="reference internal" href="implementations.snpsea.html#funkea.implementations.snpsea.SNPsea" title="funkea.implementations.snpsea.SNPsea"><code class="xref py py-class docutils literal notranslate"><span class="pre">SNPsea</span></code></a>; however, <span class="math notranslate nohighlight">\(\mathbf{L}\)</span> is constructed in a slightly different way. As
mentioned above, an annotation <span class="math notranslate nohighlight">\(j\)</span> and a locus <span class="math notranslate nohighlight">\(l\)</span> are considered “linked”, if the
lead variant or one of its LD proxies overlaps it. If the overlap is with an LD proxy, the
annotation as to also be withing 500kb of the lead variant.
Here, we define the controlling covariates <span class="math notranslate nohighlight">\(\mathbf{C}\)</span> as being binary; however, any type
of variant-level covariate is valid. In the original formulation of <code class="docutils literal notranslate"><span class="pre">GARFIELD</span></code>,
<span class="math notranslate nohighlight">\(\mathbf{C}\)</span> is divided into two feature type subsets: (1) binned distance from
transcription start site (TSS), and (2) binned number of LD proxies. Formulated differently,
<span class="math notranslate nohighlight">\(\mathbf{C}\)</span> is an indicator matrix, telling us how far from the nearest gene a given lead
variant is, and how much LD it is experiencing. The enrichment <span class="math notranslate nohighlight">\(e_i\)</span> for given partition
<span class="math notranslate nohighlight">\(i\)</span> is then defined as follows:</p>
<div class="math notranslate nohighlight">
\[e_i = \exp (\beta_i)\]</div>
<p>The significance of the enrichment is then computed in the standard way for linear models:</p>
<div class="math notranslate nohighlight">
\[p_i = 2 F \left(- \left|\frac{\log e_i}{SE(\log e_i)} \right| \right)\]</div>
<p>where <span class="math notranslate nohighlight">\(F\)</span> is the cumulative density function of a unit Gaussian.</p>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id2" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>Iotchkova, V., Ritchie, G.R., Geihs, M., Morganella, S., Min, J.L., Walter, K.,
Timpson, N.J., UK10K Consortium, Dunham, I., Birney, E. and Soranzo, N., 2019. GARFIELD
classifies disease-relevant genomic features through integration of functional
annotations with association signals. Nature genetics, 51(2), pp.343-353.</p>
</aside>
</aside>
<dl class="py method">
<dt class="sig sig-object py" id="funkea.implementations.garfield.GARFIELD.default">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">default</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotation:</span> <span class="pre">~typing.Optional[~funkea.core.data.AnnotationComponent]</span> <span class="pre">=</span> <span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params:</span> <span class="pre">~funkea.implementations.garfield.Params</span> <span class="pre">=</span> <span class="pre">Params(expansion_r2_threshold=0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ld_pruning_r2_threshold=0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">within_span=500000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">association_threshold=1e-08</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">controlling_covariates=&lt;funkea.implementations.garfield.ControllingCovariates</span> <span class="pre">object&gt;)</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Self</span></span></span><a class="headerlink" href="#funkea.implementations.garfield.GARFIELD.default" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the default workflow for <code class="docutils literal notranslate"><span class="pre">GARFIELD</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>annotation</strong> – The annotation to use for the enrichment. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the default
annotation will be used (GTEx).</p></li>
<li><p><strong>params</strong> – The parameters to use for the workflow.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The default <code class="docutils literal notranslate"><span class="pre">GARFIELD</span></code> workflow.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.garfield.Method">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.garfield.</span></span><span class="sig-name descname"><span class="pre">Method</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">partition_id_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">continuous_covariates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">categorical_covariates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">('binned_tss_distance',</span> <span class="pre">'binned_n_proxies')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">association_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1e-08</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.garfield.Method" title="Permalink to this definition">¶</a></dt>
<dd><p>Enrichment method using GARFIELD.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>partition_id_col</strong> – The name of the column containing the partition IDs.</p></li>
<li><p><strong>continuous_covariates</strong> – The names of the continuous covariates.</p></li>
<li><p><strong>categorical_covariates</strong> – The names of the categorical covariates.</p></li>
<li><p><strong>association_threshold</strong> – The threshold for the association p-value. Variants with a p-value
below this threshold will be assigned a positive label in the logistic regression.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.garfield.Params">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.garfield.</span></span><span class="sig-name descname"><span class="pre">Params</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expansion_r2_threshold:</span> <span class="pre">~funkea.core.utils.types.UnitFloat</span> <span class="pre">=</span> <span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ld_pruning_r2_threshold:</span> <span class="pre">~funkea.core.utils.types.UnitFloat</span> <span class="pre">=</span> <span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">within_span:</span> <span class="pre">~pydantic.types.NonNegativeInt</span> <span class="pre">=</span> <span class="pre">500000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">association_threshold:</span> <span class="pre">~funkea.core.utils.types.UnitFloat</span> <span class="pre">=</span> <span class="pre">1e-08</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">controlling_covariates:</span> <span class="pre">funkea.implementations.garfield.ControllingCovariates</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">&lt;funkea.implementations.garfield.ControllingCovariates</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.garfield.Params" title="Permalink to this definition">¶</a></dt>
<dd><p>Parameters for the default configuration of <a class="reference internal" href="#funkea.implementations.garfield.GARFIELD" title="funkea.implementations.garfield.GARFIELD"><code class="xref py py-class docutils literal notranslate"><span class="pre">GARFIELD</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>expansion_r2_threshold</strong> – The minimum <span class="math notranslate nohighlight">\(r^2\)</span> of a variant’s LD proxies.
(<code class="docutils literal notranslate"><span class="pre">default=0.8</span></code>)</p></li>
<li><p><strong>ld_pruning_r2_threshold</strong> – The maximum <span class="math notranslate nohighlight">\(r^2\)</span> between two variants in the GWAS
sumstats. (<code class="docutils literal notranslate"><span class="pre">default=0.01</span></code>)</p></li>
<li><p><strong>within_span</strong> – The maximum distance between a given lead variant and an overlapped
annotation. (<code class="docutils literal notranslate"><span class="pre">default=500_000</span></code>)</p></li>
<li><p><strong>association_threshold</strong> – A <span class="math notranslate nohighlight">\(p\)</span>-value threshold, where variants below this value
will be assigned the positive label in the logistic regression. (<code class="docutils literal notranslate"><span class="pre">default=1e-8</span></code>)</p></li>
<li><p><strong>controlling_covariates</strong> – The controlling covariates in the regression.
(<code class="docutils literal notranslate"><span class="pre">default=ControllingCovariates()</span></code>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.garfield.Pipeline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.garfield.</span></span><span class="sig-name descname"><span class="pre">Pipeline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus_definition:</span> <span class="pre">~funkea.components.locus_definition.LocusDefinition</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_selector:</span> <span class="pre">~typing.Optional[~funkea.components.variant_selection.VariantSelection]</span> <span class="pre">=</span> <span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">controlling_covariates:</span> <span class="pre">funkea.implementations.garfield.ControllingCovariates</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">&lt;funkea.implementations.garfield.ControllingCovariates</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.garfield.Pipeline" title="Permalink to this definition">¶</a></dt>
<dd><p>The GARFIELD data pipeline.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>locus_definition</strong> – The locus definition component.</p></li>
<li><p><strong>variant_selector</strong> – The variant selection component. (<code class="docutils literal notranslate"><span class="pre">default=None</span></code>)</p></li>
<li><p><strong>controlling_covariates</strong> – The controlling covariates component. These are the variant-level
controlling covariates used in the regression. (<code class="docutils literal notranslate"><span class="pre">default=None</span></code>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.garfield.WithinSpan">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.garfield.</span></span><span class="sig-name descname"><span class="pre">WithinSpan</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">pydantic.types.NonNegativeInt</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">pydantic.types.NonNegativeInt</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="core.data.html#funkea.core.data.AnnotationComponent" title="funkea.core.data.AnnotationComponent"><span class="pre">funkea.core.data.AnnotationComponent</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.garfield.WithinSpan" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks whether a lead variant is within a particular distance of an annotation.</p>
<p>This is an additional requirement in the original implementation of <a class="reference internal" href="#funkea.implementations.garfield.GARFIELD" title="funkea.implementations.garfield.GARFIELD"><code class="xref py py-class docutils literal notranslate"><span class="pre">GARFIELD</span></code></a>. It is
applied after the <a class="reference internal" href="components.locus_definition.html#funkea.components.locus_definition.Overlap" title="funkea.components.locus_definition.Overlap"><code class="xref py py-class docutils literal notranslate"><span class="pre">Overlap</span></code></a> step in the locus definition.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>span</strong> – The upstream and downstream distances, within which variant-annotation pairs are
retained.</p></li>
<li><p><strong>annotation</strong> – The annotation component. Held optional, such that it can be passed from
<a class="reference internal" href="components.locus_definition.html#funkea.components.locus_definition.Compose" title="funkea.components.locus_definition.Compose"><code class="xref py py-class docutils literal notranslate"><span class="pre">Compose</span></code></a> after initialisation.
(<code class="docutils literal notranslate"><span class="pre">default=None</span></code>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="funkea.implementations.garfield.fit_garfield_model">
<span class="sig-prename descclassname"><span class="pre">funkea.implementations.garfield.</span></span><span class="sig-name descname"><span class="pre">fit_garfield_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">categorical_covariates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">continuous_covariates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partition_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="headerlink" href="#funkea.implementations.garfield.fit_garfield_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit a GARFIELD model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> – The dataframe to fit the model on.</p></li>
<li><p><strong>annotation_col</strong> – The name of the annotation column.</p></li>
<li><p><strong>categorical_covariates</strong> – The names of the categorical covariates.</p></li>
<li><p><strong>continuous_covariates</strong> – The names of the continuous covariates.</p></li>
<li><p><strong>target_col</strong> – The name of the target column.</p></li>
<li><p><strong>partition_cols</strong> – The names of the partition columns.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Enrichment results produced by the model.</p>
</dd>
</dl>
</dd></dl>

</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">funkea</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../user_guide.html">User Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="components.html">funkea.components</a></li>
<li class="toctree-l2"><a class="reference internal" href="core.html">funkea.core</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="implementations.html">funkea.implementations</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">funkea API</a><ul>
  <li><a href="implementations.html">Implementations</a><ul>
      <li>Previous: <a href="implementations.fisher.html" title="previous chapter">Fisher</a></li>
      <li>Next: <a href="implementations.ldsc.html" title="next chapter">LDSC</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2023-2023, Benjamin Tenmann.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="../_sources/reference/implementations.garfield.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>