
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>DEPICT &#8212; funkea 0.1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Fisher" href="implementations.fisher.html" />
    <link rel="prev" title="Implementations" href="implementations.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="module-funkea.implementations.depict">
<span id="depict"></span><h1>DEPICT<a class="headerlink" href="#module-funkea.implementations.depict" title="Permalink to this heading">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.depict.DEPICT">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.depict.</span></span><span class="sig-name descname"><span class="pre">DEPICT</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pipeline</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Pipeline</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Method</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.depict.DEPICT" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes functional enrichment using <code class="docutils literal notranslate"><span class="pre">DEPICT</span></code>.</p>
<p><code class="docutils literal notranslate"><span class="pre">DEPICT</span></code> is a popular method for gene prioritisation, but also has proven itself useful for
functional enrichment analysis <a href="#id5"><span class="problematic" id="id1">[1]_</span></a>. Similarly to <a class="reference internal" href="implementations.snpsea.html#funkea.implementations.snpsea.SNPsea" title="funkea.implementations.snpsea.SNPsea"><code class="xref py py-class docutils literal notranslate"><span class="pre">SNPsea</span></code></a>, it uses activity matrix
<span class="math notranslate nohighlight">\(\mathbf{X} \in \mathbb{R}^{K \times |\mathcal{G}|}\)</span> to assign the annotations
<span class="math notranslate nohighlight">\(s_{ij}\)</span> to the (soft) partitions <span class="math notranslate nohighlight">\(i \in \{1, \dots, K\}\)</span>, by normalising the
activities over the <span class="math notranslate nohighlight">\(K\)</span> partitions. However, in this case, <span class="math notranslate nohighlight">\(\bar{\mathbf{X}}\)</span> is
produced from <span class="math notranslate nohighlight">\(\mathbf{X}\)</span> via <span class="math notranslate nohighlight">\(z\)</span>-scoring, rather than normalising to unit length.
Also, much like <a class="reference internal" href="implementations.snpsea.html#funkea.implementations.snpsea.SNPsea" title="funkea.implementations.snpsea.SNPsea"><code class="xref py py-class docutils literal notranslate"><span class="pre">SNPsea</span></code></a>, <code class="docutils literal notranslate"><span class="pre">DEPICT</span></code> defines bipartite adjacency matrix
<span class="math notranslate nohighlight">\(\mathbf{L} \in \{0, 1\}^{|\mathcal{G}| \times M}\)</span>, which defines whether an annotation
and a locus overlap. It then computes the scores between each partition <span class="math notranslate nohighlight">\(i\)</span> and locus as
follows:</p>
<div class="math notranslate nohighlight">
\[\mathbf{K} = \bar{\mathbf{X}} \mathbf{L} \cdot \mathrm{diag}(\mathbf{L}^T \mathbf{1})^{-1}\]</div>
<p>The loci are defined as the LD pruned genome-wide significant variants and their LD proxies.
Each locus is overlapped with the genomic annotations and loci which do not overlap any
annotations are associated with their nearest annotation. Loci which share annotations are
fused.
From <span class="math notranslate nohighlight">\(\mathbf{K}\)</span>, we then get the biased estimate of the enrichment <span class="math notranslate nohighlight">\(\hat{e}_i\)</span> of
partition <span class="math notranslate nohighlight">\(i\)</span>:</p>
<div class="math notranslate nohighlight">
\[\hat{e}_i = \frac{\mu_i}{\sigma_i}\]</div>
<p>where <span class="math notranslate nohighlight">\(\mu_i\)</span> and <span class="math notranslate nohighlight">\(\sigma_i\)</span> are the sample mean and standard deviation of the
partition-locus scores <span class="math notranslate nohighlight">\(K_{ij}\)</span>. The final bias corrected enrichment <span class="math notranslate nohighlight">\(e_i\)</span> is then
defined as</p>
<div class="math notranslate nohighlight">
\[e_i = \frac{\hat{e}_i - \mathbb{E}[\hat{e}_i]}{SE(\hat{e}_i)}\]</div>
<p>where <span class="math notranslate nohighlight">\(\mathbb{E}[\hat{e}_i]\)</span> and <span class="math notranslate nohighlight">\(SE(\hat{e}_i)\)</span> are the expectation and standard
error of <span class="math notranslate nohighlight">\(\hat{e}_i\)</span>, which are approximated empirically via Monte Carlo sampling. Since
we assume a normal distribution of <span class="math notranslate nohighlight">\(e_i\)</span>, we get <span class="math notranslate nohighlight">\(p_i = 1 - F(e_i)\)</span>, where
<span class="math notranslate nohighlight">\(F(e_i)\)</span> is the cumulative distribution function of the unit Gaussian.</p>
<p>Finally, <code class="docutils literal notranslate"><span class="pre">DEPICT</span></code> estimates the false discovery rate (FDR) for each <span class="math notranslate nohighlight">\(p_i\)</span> by running
<span class="math notranslate nohighlight">\(N\)</span> repetitions of the above steps using <span class="math notranslate nohighlight">\(M\)</span> randomly generated loci. Each generated
locus will have the same number of annotations as their “real” counterpart. The FDR is computed
as follows:</p>
<div class="math notranslate nohighlight">
\[FDR(p_i) = \frac{1}{R_i N} \sum_k^N \mathbb{1} \left[ \tilde{p}_{ik} \leq p_i \right]\]</div>
<p>where <span class="math notranslate nohighlight">\(R_i\)</span> is the ordinal rank of <span class="math notranslate nohighlight">\(p_i\)</span> and <span class="math notranslate nohighlight">\(\tilde{p}_{ik}\)</span> is the
<span class="math notranslate nohighlight">\(p\)</span>-value of the <span class="math notranslate nohighlight">\(k\)</span>-th null enrichment for partition <span class="math notranslate nohighlight">\(i\)</span>. For the sake of
consistency with the other methods, we did not include this step.</p>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id2" role="note">
<span class="label"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></span>
<p>Pers, T.H., Karjalainen, J.M., Chan, Y., Westra, H.J., Wood, A.R., Yang, J.,
Lui, J.C., Vedantam, S., Gustafsson, S., Esko, T. and Frayling, T., 2015. Biological
interpretation of genome-wide association studies using predicted gene functions.
Nature communications, 6(1), p.5890.</p>
</aside>
</aside>
<dl class="py method">
<dt class="sig sig-object py" id="funkea.implementations.depict.DEPICT.default">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">default</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="core.data.html#funkea.core.data.AnnotationComponent" title="funkea.core.data.AnnotationComponent"><span class="pre">AnnotationComponent</span></a><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#funkea.implementations.depict.Params" title="funkea.implementations.depict.Params"><span class="pre">Params</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">Params(association_threshold=5e-08,</span> <span class="pre">expansion_r2_threshold=0.5,</span> <span class="pre">ld_pruning_r2_threshold=0.1)</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Self</span></span></span><a class="headerlink" href="#funkea.implementations.depict.DEPICT.default" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the default configuration of <code class="docutils literal notranslate"><span class="pre">DEPICT</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>annotation</strong> – The annotation component to use. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the default GTEx annotation
component will be used. (<code class="docutils literal notranslate"><span class="pre">default=None</span></code>)</p></li>
<li><p><strong>params</strong> – The parameters to use. (<code class="docutils literal notranslate"><span class="pre">default=DEPICT.Params()</span></code>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The default configuration of <code class="docutils literal notranslate"><span class="pre">DEPICT</span></code>.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.depict.Method">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.depict.</span></span><span class="sig-name descname"><span class="pre">Method</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">partition_id_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.depict.Method" title="Permalink to this definition">¶</a></dt>
<dd><p>The method used to compute the enrichment score for DEPICT.</p>
<p>This method is based on the one described in the DEPICT paper <a href="#id6"><span class="problematic" id="id3">[1]_</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>partition_id_col</strong> – The name of the column containing the partition ID.</p>
</dd>
</dl>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id4" role="note">
<span class="label"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></span>
<p>Pers, T.H., Karjalainen, J.M., Chan, Y., Westra, H.J., Wood, A.R., Yang, J.,
Lui, J.C., Vedantam, S., Gustafsson, S., Esko, T. and Frayling, T., 2015. Biological
interpretation of genome-wide association studies using predicted gene functions.
Nature communications, 6(1), p.5890.</p>
</aside>
</aside>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.depict.NullLoci">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.depict.</span></span><span class="sig-name descname"><span class="pre">NullLoci</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataset</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.depict.NullLoci" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads the precomputed DEPICT null-loci.</p>
<p>The default data were taken from the original implementation. As a consequence, they use a
specific annotation type: genes. Hence, the default null loci are not applicable to annotations
which are not genes (with Ensembl IDs).
Please refer to the original paper for details on how to generate these loci for other
annotation types.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dataset</strong> – Path to the dataframe containing the null loci, or the dataframe itself.</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.depict.NullLociColumns">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.depict.</span></span><span class="sig-name descname"><span class="pre">NullLociColumns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'annotations'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'size'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_loci</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'n_loci'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.depict.NullLociColumns" title="Permalink to this definition">¶</a></dt>
<dd><p>Column names for the null loci dataframe.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.depict.OverlapAndNearestAnnotation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.depict.</span></span><span class="sig-name descname"><span class="pre">OverlapAndNearestAnnotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="core.data.html#funkea.core.data.AnnotationComponent" title="funkea.core.data.AnnotationComponent"><span class="pre">AnnotationComponent</span></a><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">join_columns</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chr'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_variants</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.depict.OverlapAndNearestAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Overlaps annotations, or nearest annotation if not overlaps were found.</p>
<p>This is a DEPICT-specific locus definition step, which says that if a given locus has no
overlapping annotation, we should include the nearest annotation.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.depict.Params">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.depict.</span></span><span class="sig-name descname"><span class="pre">Params</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">association_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="core.utils.types.html#funkea.core.utils.types.UnitFloat" title="funkea.core.utils.types.UnitFloat"><span class="pre">UnitFloat</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5e-08</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expansion_r2_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="core.utils.types.html#funkea.core.utils.types.UnitFloat" title="funkea.core.utils.types.UnitFloat"><span class="pre">UnitFloat</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ld_pruning_r2_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="core.utils.types.html#funkea.core.utils.types.UnitFloat" title="funkea.core.utils.types.UnitFloat"><span class="pre">UnitFloat</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.depict.Params" title="Permalink to this definition">¶</a></dt>
<dd><p>Parameters for the default configuration of <code class="docutils literal notranslate"><span class="pre">DEPICT</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>association_threshold</strong> – The <span class="math notranslate nohighlight">\(p\)</span>-value threshold used to select genome-wide significant
variants. (<code class="docutils literal notranslate"><span class="pre">default=5e-8</span></code>)</p></li>
<li><p><strong>expansion_r2_threshold</strong> – The minimum <span class="math notranslate nohighlight">\(r^2\)</span> value for a variant-LD proxy pair.
(<code class="docutils literal notranslate"><span class="pre">default=0.5</span></code>)</p></li>
<li><p><strong>ld_pruning_r2_threshold</strong> – The maximum <span class="math notranslate nohighlight">\(r^2\)</span> between any pair of variants in the
provided sumstats. (<code class="docutils literal notranslate"><span class="pre">default=0.1</span></code>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="funkea.implementations.depict.Pipeline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">funkea.implementations.depict.</span></span><span class="sig-name descname"><span class="pre">Pipeline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus_definition:</span> <span class="pre">~funkea.components.locus_definition.LocusDefinition</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_selector:</span> <span class="pre">~typing.Optional[~funkea.components.variant_selection.VariantSelection]</span> <span class="pre">=</span> <span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">null_loci:</span> <span class="pre">~funkea.implementations.depict.NullLoci</span> <span class="pre">=</span> <span class="pre">&lt;funkea.implementations.depict.NullLoci</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_permutations:</span> <span class="pre">~pydantic.types.NonNegativeInt</span> <span class="pre">=</span> <span class="pre">500</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funkea.implementations.depict.Pipeline" title="Permalink to this definition">¶</a></dt>
<dd><p>Data pipeline for running DEPICT.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>locus_definition</strong> – The locus definition.</p></li>
<li><p><strong>variant_selector</strong> – The variant selector.</p></li>
<li><p><strong>null_loci</strong> – The null loci dataset.</p></li>
<li><p><strong>n_permutations</strong> – The number of permutations to generate for the significance test.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="funkea.implementations.depict.Pipeline.get_matched_null_loci">
<span class="sig-name descname"><span class="pre">get_matched_null_loci</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">loci</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="headerlink" href="#funkea.implementations.depict.Pipeline.get_matched_null_loci" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the matched null loci for the given real loci.</p>
<p>Matching here is done by number of annotations in the locus. I.e. we find the null loci
which have the closest number of annotations to the real loci.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>loci</strong> – The real loci.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The matched null loci.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">funkea</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../user_guide.html">User Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="components.html">funkea.components</a></li>
<li class="toctree-l2"><a class="reference internal" href="core.html">funkea.core</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="implementations.html">funkea.implementations</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">funkea API</a><ul>
  <li><a href="implementations.html">Implementations</a><ul>
      <li>Previous: <a href="implementations.html" title="previous chapter">Implementations</a></li>
      <li>Next: <a href="implementations.fisher.html" title="next chapter">Fisher</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2023-2023, Benjamin Tenmann.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="../_sources/reference/implementations.depict.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>